### ISSUE TITLE
Friend list / block list

### ISSUE THREAD

**Issue by Ashoat** [Ashoat/squadcal#12]
*Right now*: everybody is "friends" with everybody. Anybody can start a thread with anybody. You don't have a friend list - there is just a list of threads, each of which contains one or more users.

*In the future*: when you search for a user, or select them from the UI, the app pulls up your thread with that user. This thread is visible even if you have the user blocked. When you open the thread:
- If the user has blocked you or you have blocked the user, you'll be able to see your past messages, but won't be able to send any new ones
- Otherwise, if neither user has blocked the other, they should be able to chat, even if they are not friends
- If the user isn't your friend yet, there should be a prompt to friend request them
- If the user has friend requested you, there should be a prompt to respond to the request

In the "More" tab there should be a way to see your friend list and your block list, and to remove or add users to either.

~~In the "Compose thread" screen you should be able to create a thread with users you're not currently friends with.~~ However, you shouldn't be able to create a thread with a user who has blocked you, or a user you have blocked. ~~If you create a thread with somebody you're not currently friends with, the behavior should be the same as if you search for a user (as described below).~~

If you search for a user in the "thread list" page:
- Entries for friends should list your existing thread with them
- Entries for non-friends should list a fake thread. When you click through, it gives you the option to friend request or send a message to the user, which will start the thread

**Comment by Ashoat** [comment]
Realized I never mentioned this: once we get the friend list / block list in place, we should put some permission checks in place so that blocked users can't message each other (only for 2 person threads) and that you need to be friends with somebody in order to add them to a thread.

**Comment by Ashoat** [comment]
Things we haven't done yet:

1. Reconsider whether `RelationshipListItem` should have a single-item tooltip for some actions, and a button for other actions
2. Sort the `RelationshipList`, and put pending items (eg. not accepted friend requests) at the top
3. Update server to check permissions (see [previous comment](https://github.com/Ashoat/squadcal/issues/12#issuecomment-656665511))
4. Have server set permissions of any 1-on-1 chat threads when blocks occur. Make sure client prevents messaging or being messaged by blocked people, but still allow people to see past messages
5. Show a prompt to friend somebody on any 1-on-1 chat threads with people you haven't friended or blocked yet (and haven't blocked you). If they have friend requested you already, show a prompt to accept the friend request
6. Have friend requests create a thread with somebody
7. Show fake threads for non-friends in `ThreadList` search results
8. Have client stop listing non-friends as options for adding to thread (or creating new thread)

**Comment by Ashoat** [comment]
Realized another thing I hadn't mentioned: friend request actions (sending a friend request and accepting a friend request, but not rejecting a friend request or blocks) should create messages in the 1-on-1 chat thread. We'll need to create new message types to make this work. Adding this to the list above

**Comment by Ashoat** [comment]
Talked with Tomek about the requirements a bit yesterday and wanted to document it. He pointed out that this issue had some conflicting info about whether or not a user can add a non-friend to a thread.

We decided to stick with the [July 10th](https://github.com/Ashoat/squadcal/issues/12#issuecomment-656665511) comment I made above:

> you need to be friends with somebody in order to add them to a thread. We'll also allow people to add each other to subthreads if they're in the parent thread together, even if they're not friends

We talked a bit about how that felt inconsistent with fake threads, where it's possible for non-friends to communicate with each other. Given this distinction we decided it made sense to introduce a new kind of threads: personal threads.

We also talked about how it may be unclear to a user why they can't add a non-friend to a thread. Instead of simply not listing non-friends as options, we could list them but make them unselectable and say "not a friend" or something. Maybe they would pop up an `Alert` when you press them. Ideally they won't show up until you type the first character of their username into the `TagInput`.

**Comment by Ashoat** [comment]
Kasia brought something up over chat:

> I also found one thing that I'm not sure is done on purpose - when two users block each other, there is no option for any of them to unblock. If user A  tries to add B to friends, there is alert with 'unknown error, try again', then B unblocked, and the B can unblock A by clicking on a button. A can try to add B to friends, and each time there's the same alert - I think maybe there should be some kind of explanation, user B blocked you or something like that?

I think there are two issues here:

1. It should always be possible to unblock a user that you have blocked. The UI should list an option for this and the server should not error.
2. It should not be possible to send a friend request to a user who has blocked you. For the friend list, `RelationshipUpdateModal` should not list users who have blocked you. (For the block list, it should still list them.) The server should also prevent a friend request from being sent to a user who has blocked you.

I've added these to the list of remaining tasks above.

**Comment by Ashoat** [comment]
After thinking about the permissions logic at today's weekly meeting, we decided to split that work off into a new issue: #116

**Comment by Ashoat** [comment]
One more thing I thought of. I'd like to clarify a couple of situations where the viewer might not be able to do something because they've blocked a particular user:

1. If the viewer doesn't have `VOICED` permissions in a personal thread because they blocked the target, let's replace the standard "You don't have permission to send messages." notice with "You can't send messages to a user that you've blocked." We'll want to do this on both native and web, so we should probably extract the notice-determining logic into a function somewhere in lib.
2. On native, let's handle the case where the viewer is trying to add a user they've blocked to a thread, through `AddUsersModal` or `ComposeThread`. We won't show users that you've blocked if the `searchText` is empty, but if the viewer searches for a blocked user, we'll display that user at the bottom of the list, with a `notice` of `"you've blocked this user"` and an `alertText` with some more details.

**Comment by Ashoat** [comment]
I think all of the work in this issues is either complete, or has been split out into separate issues. Great work team!
